Proxy logging stack traces when shutting down

```
ERROR  [2025-04-10T15:45:22.683Z] [forward - xxxxxxx 2] stroom.proxy.repo.ParallelExecutor -  Error running parallel executor 'forward - xxxxxx' task: UncheckedInterruptedException
stroom.util.concurrent.UncheckedInterruptedException: null
        at stroom.util.concurrent.UncheckedInterruptedException.create(UncheckedInterruptedException.java:43)
        at stroom.util.concurrent.UncheckedInterruptedException.create(UncheckedInterruptedException.java:26)
        at stroom.proxy.app.handler.DirQueue.next(DirQueue.java:102)
        at stroom.proxy.app.handler.DirQueueTransfer.run(DirQueueTransfer.java:20)
        at stroom.proxy.repo.ParallelExecutor.runTask(ParallelExecutor.java:199)
        at stroom.proxy.repo.ParallelExecutor.run(ParallelExecutor.java:159)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.InterruptedException: null
        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1722)
        at stroom.proxy.app.handler.DirQueue.next(DirQueue.java:85)
```

Interruption is expected during shutdown so it should not log the stack

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/DirQueue.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/DirQueue.java
@@ -4,6 +4,7 @@ import stroom.proxy.repo.queue.QueueMonitor;
 import stroom.proxy.repo.queue.QueueMonitors;
 import stroom.proxy.repo.store.FileStores;
 import stroom.util.concurrent.UncheckedInterruptedException;
+import stroom.util.exception.ThrowingSupplier;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
 import stroom.util.logging.LogUtil;
@@ -152,13 +153,21 @@ public class DirQueue {
                 // Increment the sequence id.
                 final long id = ++writeId;
                 queueMonitor.setWritePos(id);
+                final Path targetDir = DirUtil.createPath(rootDir, id);
+                final Path targetParent = targetDir.getParent();
                 try {
-                    final Path targetDir = DirUtil.createPath(rootDir, id);
-                    DirUtil.ensureDirExists(targetDir.getParent());
+                    DirUtil.ensureDirExists(targetParent);
                     Files.move(sourceDir, targetDir, StandardCopyOption.ATOMIC_MOVE);
                     LOGGER.trace("{} ({}) - Added sourceDir {}", name, rootDir, sourceDir);
                 } catch (final IOException e) {
-                    LOGGER.error(e::getMessage, e);
+                    final boolean targetParentExists = LogUtil.swallowExceptions(
+                                    ThrowingSupplier.unchecked(() -> Files.exists(targetParent)))
+                            .orElse(false);
+                    final boolean sourceExists = LogUtil.swallowExceptions(
+                                    ThrowingSupplier.unchecked(() -> Files.exists(sourceDir)))
+                            .orElse(false);
+                    LOGGER.error("Error moving {} -> {}, sourceExists: {}, targetParentExists: {}, msg: {}",
+                            sourceDir, targetDir, sourceExists, targetParentExists, LogUtil.exceptionMessage(e), e);
                     throw new UncheckedIOException(e);
                 }
                 condition.signalAll();

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/HttpSender.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/HttpSender.java
@@ -309,8 +309,13 @@ public class HttpSender implements StreamDestination {
             final ResponseStatus responseStatus = checkConnectionResponse(response, attributeMap);
             final StroomStatusCode stroomStatusCode = responseStatus.stroomStatusCode;
             final String receiptId = responseStatus.receiptId;
-            LOGGER.debug("'{}' - stroomStatusCode: {}, receiptId {}, contentLength: {}",
-                    forwarderName, stroomStatusCode, receiptId, contentLength);
+            LOGGER.debug(() -> LogUtil.message(
+                    "'{}' - stroomStatusCode: {}, receiptId {}, contentLength: {}, compression: '{}'",
+                    forwarderName,
+                    stroomStatusCode,
+                    receiptId,
+                    ByteSize.ofBytes(contentLength),
+                    attributeMap.get(StandardHeaderArguments.COMPRESSION)));
 
             final EventType eventType = stroomStatusCode == StroomStatusCode.OK
                     ? EventType.SEND
@@ -481,7 +486,7 @@ public class HttpSender implements StreamDestination {
 
         try (final InputStream inputStream = response.getEntity().getContent()) {
             if (inputStream != null) {
-                return IOUtils.toString(inputStream, StandardCharsets.UTF_8);
+                return NullSafe.trim(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
             }
         } catch (final IOException ioex) {
             LOGGER.debug(ioex.getMessage(), ioex);

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/PreAggregator.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/PreAggregator.java
@@ -38,6 +38,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
@@ -452,15 +453,30 @@ public class PreAggregator {
         return Collections.singletonList(new Part(partItems, partBytes, List.copyOf(partEntries)));
     }
 
-    private void closeAggregate(final FeedKey feedKey,
-                                final AggregateState aggregateState) {
-        LOGGER.debug("Closing aggregate: {}", aggregateState);
+    private boolean closeAggregate(final FeedKey feedKey,
+                                   final AggregateState aggregateState) {
+        LOGGER.debug(() -> LogUtil.message("closeAggregate() - feedKey: {}, {}, waiting for lock",
+                feedKey, aggregateState));
         final Lock lock = feedKeyLock.get(feedKey);
         lock.lock();
         try {
-            destination.accept(aggregateState.aggregateDir);
-            aggregateStateMap.remove(feedKey);
-            LOGGER.debug("Closed aggregate: {}", aggregateState);
+            // Now we hold the feedKey lock, re-check the aggregateStateMap
+            if (aggregateStateMap.containsKey(feedKey)) {
+                LOGGER.debug(() -> LogUtil.message("closeAggregate() - feedKey: {}, {}, acquired lock",
+                        feedKey, aggregateState));
+
+                destination.accept(aggregateState.aggregateDir);
+                aggregateStateMap.remove(feedKey);
+                LOGGER.debug(() -> LogUtil.message("closeAggregate() - feedKey: {}, {}, closed aggregate",
+                        feedKey, aggregateState));
+                return true;
+            } else {
+                LOGGER.debug(() -> LogUtil.message(
+                        "closeAggregate() - feedKey: {}, {}, " +
+                        "feedKey not in aggregateStateMap, another thread must have closed it.",
+                        feedKey, aggregateState));
+                return false;
+            }
         } finally {
             lock.unlock();
         }
@@ -570,13 +586,23 @@ public class PreAggregator {
      */
     private synchronized void closeOldAggregates() {
         final AtomicInteger count = new AtomicInteger();
-        aggregateStateMap.forEach((feedKey, aggregateState) -> {
-            if (aggregateState.isAggregateTooOld(aggregatorConfig)) {
-                // Close the current aggregate.
-                count.incrementAndGet();
-                closeAggregate(feedKey, aggregateState);
+        final Set<FeedKey> feedKeys = aggregateStateMap.keySet();
+        for (final FeedKey feedKey : feedKeys) {
+            // It's possible another thread may have removed it
+            final AggregateState aggregateState = aggregateStateMap.get(feedKey);
+            if (aggregateState != null
+                && aggregateState.isAggregateTooOld(aggregatorConfig)) {
+                // Close the current aggregate, under a feedKey lock, so again,
+                // another thread may beat us
+                final boolean didClose = closeAggregate(feedKey, aggregateState);
+                if (didClose) {
+                    count.incrementAndGet();
+                } else {
+                    LOGGER.debug("closeAggregate() - feedKey: {}, aggregateState: {}, didn't close",
+                            feedKey, aggregateState);
+                }
             }
-        });
+        }
         if (LOGGER.isDebugEnabled()) {
             if (count.get() > 0) {
                 LOGGER.debug("closeOldAggregates() - closed {} old aggregates", count);

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ProxyRequestHandler.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ProxyRequestHandler.java
@@ -8,12 +8,14 @@ import stroom.receive.common.ReceiptIdGenerator;
 import stroom.receive.common.RequestAuthenticator;
 import stroom.receive.common.RequestHandler;
 import stroom.receive.common.StroomStreamException;
+import stroom.util.NullSafe;
 import stroom.util.cert.CertificateExtractor;
 import stroom.util.concurrent.UniqueId;
 import stroom.util.date.DateUtil;
 import stroom.util.io.StreamUtil;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
+import stroom.util.logging.LogUtil;
 import stroom.util.net.HostNameUtil;
 
 import jakarta.inject.Inject;
@@ -23,6 +25,7 @@ import org.apache.hc.core5.http.HttpStatus;
 
 import java.io.IOException;
 import java.io.PrintWriter;
+import java.time.Duration;
 import java.time.Instant;
 
 /**
@@ -82,26 +85,33 @@ public class ProxyRequestHandler implements RequestHandler {
             appendReceivedPath(attributeMap);
 
             // Treat differently depending on compression type.
-            String compression = attributeMap.get(StandardHeaderArguments.COMPRESSION);
-            if (compression != null && !compression.isEmpty()) {
-                compression = compression.toUpperCase(StreamUtil.DEFAULT_LOCALE);
-                if (!StandardHeaderArguments.VALID_COMPRESSION_SET.contains(compression)) {
-                    throw new StroomStreamException(
-                            StroomStatusCode.UNKNOWN_COMPRESSION, attributeMap, compression);
-                }
+            String compression = NullSafe.getOrElse(
+                    attributeMap.get(StandardHeaderArguments.COMPRESSION),
+                    str -> str.toUpperCase(StreamUtil.DEFAULT_LOCALE),
+                    "");
+
+            if (!StandardHeaderArguments.VALID_COMPRESSION_SET.contains(compression)) {
+                throw new StroomStreamException(
+                        StroomStatusCode.UNKNOWN_COMPRESSION, attributeMap, compression);
             }
 
+            final Receiver receiver;
             if (ZERO_CONTENT.equals(attributeMap.get(StandardHeaderArguments.CONTENT_LENGTH))) {
                 LOGGER.warn("process() - Skipping Zero Content " + attributeMap);
-
+                receiver = null;
             } else {
-                final Receiver receiver = receiverFactory.get(attributeMap);
+                receiver = receiverFactory.get(attributeMap);
                 receiver.receive(startTime, attributeMap, request.getRequestURI(), request::getInputStream);
             }
 
             response.setStatus(HttpStatus.SC_OK);
 
-            LOGGER.debug(() -> "Writing proxy receipt id attribute to response: " + receiptIdStr);
+            LOGGER.debug(() -> LogUtil.message(
+                    "Writing proxy receipt id {} to response. Receiver: {}, duration: {}, compression: '{}'",
+                    receiptIdStr,
+                    NullSafe.get(receiver, Object::getClass, Class::getSimpleName),
+                    Duration.between(startTime, Instant.now()),
+                    compression));
             try (final PrintWriter writer = response.getWriter()) {
                 writer.println(receiptIdStr);
             } catch (final IOException e) {

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ProxyRequestHandler.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ProxyRequestHandler.java
@@ -27,6 +27,7 @@ import java.io.IOException;
 import java.io.PrintWriter;
 import java.time.Duration;
 import java.time.Instant;
+import java.util.Objects;
 
 /**
  * Main entry point to handling proxy requests.
@@ -73,7 +74,6 @@ public class ProxyRequestHandler implements RequestHandler {
             requestAuthenticator.authenticate(request, attributeMap);
 
             final String receiptIdStr = receiptId.toString();
-            LOGGER.debug("Adding meta attribute {}: {}", StandardHeaderArguments.RECEIPT_ID, receiptIdStr);
             attributeMap.put(StandardHeaderArguments.RECEIPT_ID, receiptIdStr);
             attributeMap.appendItem(StandardHeaderArguments.RECEIPT_ID_PATH, receiptIdStr);
 
@@ -85,12 +85,23 @@ public class ProxyRequestHandler implements RequestHandler {
             appendReceivedPath(attributeMap);
 
             // Treat differently depending on compression type.
+            // AttributeMap values are ready trimmed
             String compression = NullSafe.getOrElse(
                     attributeMap.get(StandardHeaderArguments.COMPRESSION),
                     str -> str.toUpperCase(StreamUtil.DEFAULT_LOCALE),
                     "");
 
-            if (!StandardHeaderArguments.VALID_COMPRESSION_SET.contains(compression)) {
+            LOGGER.debug(() -> LogUtil.message(
+                    "handle() - requestUri: {}, remoteHost/Addr: {}, attributeMap: {}, ",
+                    request.getRequestURI(),
+                    Objects.requireNonNullElseGet(
+                            request.getRemoteHost(),
+                            request::getRemoteAddr),
+                    attributeMap));
+
+            if (!compression.isEmpty()
+                && !StandardHeaderArguments.VALID_COMPRESSION_SET.contains(compression)) {
+
                 throw new StroomStreamException(
                         StroomStatusCode.UNKNOWN_COMPRESSION, attributeMap, compression);
             }

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/RemoteFeedStatusService.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/RemoteFeedStatusService.java
@@ -11,6 +11,7 @@ import stroom.util.NullSafe;
 import stroom.util.cache.CacheConfig;
 import stroom.util.jersey.JerseyClientFactory;
 import stroom.util.jersey.JerseyClientName;
+import stroom.util.logging.DurationTimer;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
 import stroom.util.logging.LogUtil;
@@ -170,17 +171,26 @@ public class RemoteFeedStatusService implements FeedStatusService, HasHealthChec
             final FeedStatusConfig feedStatusConfig,
             final Function<Response, GetFeedStatusResponse> responseConsumer) {
 
-        LOGGER.debug("Sending request {}", request);
+        LOGGER.debug(() -> LogUtil.message("Sending {} to {}{}",
+                request,
+                feedStatusConfig.getFeedStatusUrl(),
+                GET_FEED_STATUS_PATH));
 
         final WebTarget webTarget = getFeedStatusWebTarget(feedStatusConfig);
 
+        final DurationTimer timer = DurationTimer.start();
         try (final Response response = getFeedStatusResponse(feedStatusConfig, webTarget, request)) {
-            LOGGER.debug("Received response {}", response);
+            LOGGER.debug("Received response {}, duration: {}", response, timer);
             return responseConsumer.apply(response);
         } catch (Exception e) {
             throw new RuntimeException(LogUtil.message(
-                    "Error sending request {} to {}{}: {}",
-                    request, feedStatusConfig.getFeedStatusUrl(), GET_FEED_STATUS_PATH, e.getMessage()), e);
+                    "Error sending {} to {}{}, duration: {}, msg: {}",
+                    request,
+                    feedStatusConfig.getFeedStatusUrl(),
+                    GET_FEED_STATUS_PATH,
+                    timer,
+                    e.getMessage()),
+                    e);
         }
     }
 

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ZipReceiver.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ZipReceiver.java
@@ -13,13 +13,18 @@ import stroom.proxy.repo.LogStream.EventType;
 import stroom.receive.common.AttributeMapFilter;
 import stroom.receive.common.StroomStreamException;
 import stroom.util.io.ByteCountInputStream;
+import stroom.util.io.ByteSize;
 import stroom.util.io.FileName;
 import stroom.util.io.FileUtil;
+import stroom.util.logging.DurationTimer;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
+import stroom.util.logging.LogUtil;
+import stroom.util.zip.ZipUtil;
 
 import jakarta.inject.Inject;
 import jakarta.inject.Singleton;
+import org.apache.commons.compress.archivers.ArchiveEntry;
 import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
 import org.apache.commons.compress.archivers.zip.ZipArchiveInputStream;
 import org.apache.commons.compress.archivers.zip.ZipFile;
@@ -40,6 +45,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 
@@ -301,8 +307,9 @@ public class ZipReceiver implements Receiver {
                                           final Path zipFilePath,
                                           final byte[] buffer) throws IOException {
         // TODO : Worry about memory usage here storing potentially 1000's of data entries and groups.
-        LOGGER.debug("receiveZipStream() - defaultFeedName: '{}', defaultTypeName: '{}', zipFilePath: {}",
+        LOGGER.debug("receiveZipStream() - START defaultFeedName: '{}', defaultTypeName: '{}', zipFilePath: {}",
                 defaultFeedName, defaultTypeName, zipFilePath);
+        final DurationTimer timer = LogUtil.startTimerIfDebugEnabled(LOGGER);
         final List<Entry> dataEntries = new ArrayList<>();
         final Map<String, ZipEntryGroup> groups = new HashMap<>();
         final ProxyZipValidator validator = new ProxyZipValidator();
@@ -311,7 +318,7 @@ public class ZipReceiver implements Receiver {
         // Receive data.
         try (final ByteCountInputStream byteCountInputStream = new ByteCountInputStream(inputStream)) {
             try (final ZipArchiveInputStream zipInputStream = new ZipArchiveInputStream(byteCountInputStream)) {
-                // Write the incoming zip data to temp and record info about the entries as we go.
+                // Write the incoming zip data to the receiving dir and record info about the entries as we go.
                 try (final ZipWriter zipWriter = new ZipWriter(zipFilePath, buffer)) {
 
                     ZipArchiveEntry entry = zipInputStream.getNextEntry();
@@ -362,6 +369,7 @@ public class ZipReceiver implements Receiver {
                                 }
 
                                 final long size = zipWriter.writeStream(entryName, zipInputStream);
+//                                final long size = writeEntry(zipWriter, entry, zipInputStream);
                                 zipEntryGroup.setContextEntry(new ZipEntryGroup.Entry(entryName, size));
 
                             } else if (StroomZipFileType.MANIFEST.equals(stroomZipFileType)) {
@@ -372,10 +380,12 @@ public class ZipReceiver implements Receiver {
                                 }
 
                                 final long size = zipWriter.writeStream(entryName, zipInputStream);
+//                                final long size = writeEntry(zipWriter, entry, zipInputStream);
                                 zipEntryGroup.setManifestEntry(new ZipEntryGroup.Entry(entryName, size));
 
                             } else {
                                 final long size = zipWriter.writeStream(entryName, zipInputStream);
+//                                final long size = writeEntry(zipWriter, entry, zipInputStream);
                                 dataEntries.add(new ZipEntryGroup.Entry(entryName, size));
                             }
 
@@ -453,7 +463,41 @@ public class ZipReceiver implements Receiver {
             LOGGER.debug(validator.getErrorMessage());
         }
 
+        LOGGER.debug(() -> LogUtil.message(
+                "receiveZipStream() - FINISH defaultFeedName: '{}', defaultTypeName: '{}', zipFilePath: {}, " +
+                "feedKey count: {}, total entry count: {}, duration: {}",
+                defaultFeedName,
+                defaultTypeName,
+                zipFilePath,
+                feedGroups.size(),
+                LogUtil.swallowExceptions(() -> feedGroups.values().stream().mapToInt(List::size).sum())
+                        .orElse(-1),
+                timer));
+
         return new ReceiveResult(feedGroups, receivedBytes, validator.isValid());
+    }
+
+    private static long writeEntry(final ZipWriter zipWriter,
+                                   final ZipArchiveEntry sourceEntry,
+                                   final ZipArchiveInputStream zipArchiveInputStream) throws IOException {
+        Objects.requireNonNull(sourceEntry);
+        final boolean hasKnownSize = ZipUtil.hasKnownUncompressedSize(sourceEntry);
+        final long size;
+        if (ZipUtil.hasKnownUncompressedSize(sourceEntry)) {
+            // We know the size so can just write the raw entry without having to de-compress/compress it
+            zipWriter.writeRawStream(sourceEntry, zipArchiveInputStream);
+            size = sourceEntry.getSize();
+        } else {
+            // We don't know the uncompressed size, so have to effectively de-compress/compress it to find out
+            size = zipWriter.writeStream(sourceEntry.getName(), zipArchiveInputStream);
+        }
+        LOGGER.debug(() -> LogUtil.message("writeEntry() - sourceEntry: {}, hasKnownSize: {}, size: {}",
+                sourceEntry,
+                hasKnownSize,
+                (size != ArchiveEntry.SIZE_UNKNOWN
+                        ? ByteSize.ofBytes(size).toString()
+                        : "?")));
+        return size;
     }
 
     static List<Path> splitZip(final Path zipFilePath,
@@ -461,8 +505,9 @@ public class ZipReceiver implements Receiver {
                                final Map<FeedKey, List<ZipEntryGroup>> feedGroups,
                                final Path outputParentDir,
                                final byte[] buffer) throws IOException {
-        LOGGER.debug("splitZip() - zipFilePath: {}, outputParentDir: {}", zipFilePath, outputParentDir);
+        LOGGER.debug("splitZip() - START zipFilePath: {}, outputParentDir: {}", zipFilePath, outputParentDir);
         final List<Path> groupDirs = new ArrayList<>();
+        final DurationTimer timer = LogUtil.startTimerIfDebugEnabled(LOGGER);
         long id = 1;
         try (final ZipFile zipFile = createZipFile(zipFilePath)) {
             for (final Map.Entry<FeedKey, List<ZipEntryGroup>> entry : feedGroups.entrySet()) {
@@ -477,6 +522,10 @@ public class ZipReceiver implements Receiver {
                 writeZip(zipFile, zipEntryGroups, feedKey, attributeMap, fileGroup, buffer);
             }
         }
+        LOGGER.debug(() -> LogUtil.message(
+                "splitZip() - FINISH zipFilePath: {}, outputParentDir: {}, groupDirs count: {}, " +
+                "feedGroups count: {}, duration: {}",
+                zipFilePath, outputParentDir, groupDirs.size(), feedGroups.size(), timer));
         return groupDirs;
     }
 
@@ -492,16 +541,17 @@ public class ZipReceiver implements Receiver {
                                  final AttributeMap attributeMap,
                                  final FileGroup fileGroup,
                                  final byte[] buffer) throws IOException {
-        LOGGER.debug("writeZip() - feedKey: {}, fileGroup: {}", feedKey, fileGroup);
+        LOGGER.debug("writeZip() - START feedKey: {}, fileGroup: {}", feedKey, fileGroup);
+        final DurationTimer timer = LogUtil.startTimerIfDebugEnabled(LOGGER);
         try {
             // Write zip.
-            int count = 1;
+            final AtomicInteger count = new AtomicInteger(0);
             final Path entriesFile = fileGroup.getEntries();
             try (final Writer entryWriter = Files.newBufferedWriter(entriesFile)) {
                 try (final ProxyZipWriter zipWriter = new ProxyZipWriter(fileGroup.getZip(), buffer)) {
                     for (final ZipEntryGroup zipEntryGroupIn : zipEntryGroupsIn) {
                         final ZipEntryGroup zipEntryGroupOut = new ZipEntryGroup(feedKey.feed(), feedKey.type());
-                        final String baseNameOut = NumericFileNameUtil.create(count);
+                        final String baseNameOut = NumericFileNameUtil.create(count.incrementAndGet());
                         LOGGER.trace("feedKey: {}, baseNameOut: {}, zipEntryGroupIn: {} => zipEntryGroupOut: {}",
                                 feedKey, baseNameOut, zipEntryGroupIn, zipEntryGroupOut);
                         zipEntryGroupOut.setManifestEntry(
@@ -530,8 +580,6 @@ public class ZipReceiver implements Receiver {
                                         baseNameOut,
                                         StroomZipFileType.DATA));
 
-                        count++;
-
                         // Write zip entry.
                         LOGGER.trace("Writing entries file {}", entriesFile);
                         zipEntryGroupOut.write(entryWriter);
@@ -542,6 +590,9 @@ public class ZipReceiver implements Receiver {
             // Write meta.
             AttributeMapUtil.addFeedAndType(attributeMap, feedKey.feed(), feedKey.type());
             AttributeMapUtil.write(attributeMap, fileGroup.getMeta());
+            LOGGER.debug(() -> LogUtil.message(
+                    "writeZip() - FINISH feedKey: {}, fileGroup: {}, count: {}, duration: {}",
+                    feedKey, fileGroup, count, timer));
 
         } catch (final IOException e) {
             LOGGER.error(e::getMessage, e);
@@ -557,7 +608,9 @@ public class ZipReceiver implements Receiver {
         if (entry != null) {
             final ZipArchiveEntry zipEntry = zip.getEntry(entry.getName());
             final InputStream inputStream = zip.getInputStream(zipEntry);
+//            final InputStream inputStream = zip.getRawInputStream(zipEntry);
             final String outEntryName = baseNameOut + stroomZipFileType.getDotExtension();
+//            zipWriter.writeRawStream(zipEntry, outEntryName, inputStream);
             zipWriter.writeStream(outEntryName, inputStream);
             return new Entry(outEntryName, entry.getUncompressedSize());
         }

--- a/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ZipWriter.java
+++ b/stroom-proxy/stroom-proxy-app/src/main/java/stroom/proxy/app/handler/ZipWriter.java
@@ -2,6 +2,11 @@ package stroom.proxy.app.handler;
 
 import stroom.meta.api.AttributeMap;
 import stroom.meta.api.AttributeMapUtil;
+import stroom.util.io.ByteSize;
+import stroom.util.logging.DurationTimer;
+import stroom.util.logging.LambdaLogger;
+import stroom.util.logging.LambdaLoggerFactory;
+import stroom.util.logging.LogUtil;
 import stroom.util.zip.ZipUtil;
 
 import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
@@ -15,24 +20,38 @@ import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.Objects;
 
 public class ZipWriter implements AutoCloseable {
 
+    private static final LambdaLogger LOGGER = LambdaLoggerFactory.getLogger(ZipWriter.class);
+
     private final ZipArchiveOutputStream zipArchiveOutputStream;
     private final byte[] buffer;
+    /**
+     * Here to aid logging only, may be null
+     */
+    private final Path path;
 
     public ZipWriter(final Path path, final byte[] buffer) throws IOException {
-        this(new BufferedOutputStream(Files.newOutputStream(path)), buffer);
+        this(ZipUtil.createOutputStream(new BufferedOutputStream(Files.newOutputStream(path))), buffer, path);
     }
 
     public ZipWriter(final OutputStream outputStream, final byte[] buffer) {
-        this(ZipUtil.createOutputStream(outputStream), buffer);
+        this(ZipUtil.createOutputStream(outputStream), buffer, null);
     }
 
     public ZipWriter(final ZipArchiveOutputStream zipArchiveOutputStream,
                      final byte[] buffer) {
+        this(zipArchiveOutputStream, buffer, null);
+    }
+
+    private ZipWriter(final ZipArchiveOutputStream zipArchiveOutputStream,
+                      final byte[] buffer,
+                      final Path path) {
         this.zipArchiveOutputStream = zipArchiveOutputStream;
         this.buffer = buffer;
+        this.path = path;
     }
 
     public void writeDir(final String name) throws IOException {
@@ -64,12 +83,62 @@ public class ZipWriter implements AutoCloseable {
 
     private long writeStream(final ZipArchiveEntry zipArchiveEntry,
                              final InputStream inputStream) throws IOException {
+        final DurationTimer timer = LogUtil.startTimerIfDebugEnabled(LOGGER);
         putArchiveEntry(zipArchiveEntry);
         try {
-            return TransferUtil.transfer(inputStream, zipArchiveOutputStream, buffer);
+            final long bytes = TransferUtil.transfer(inputStream, zipArchiveOutputStream, buffer);
+            LOGGER.debug(() -> LogUtil.message(
+                    "writeStream() - path: {}, zipArchiveEntry: {}, bytes: {}, duration: {}",
+                    path,
+                    zipArchiveEntry,
+                    ByteSize.ofBytes(bytes),
+                    timer));
+            return bytes;
         } finally {
             closeArchiveEntry();
         }
+
+    }
+
+    public void writeRawStream(final ZipArchiveEntry sourceZipArchiveEntry,
+                               final InputStream inputStream) throws IOException {
+        writeRawStream(sourceZipArchiveEntry,
+                new ZipArchiveEntry(sourceZipArchiveEntry.getName()),
+                inputStream);
+    }
+
+    public void writeRawStream(final ZipArchiveEntry sourceZipArchiveEntry,
+                               final String destEntryName,
+                               final InputStream inputStream) throws IOException {
+        writeRawStream(sourceZipArchiveEntry,
+                new ZipArchiveEntry(destEntryName),
+                inputStream);
+    }
+
+    public void writeRawStream(final ZipArchiveEntry sourceZipArchiveEntry,
+                               final ZipArchiveEntry destZipArchiveEntry,
+                               final InputStream inputStream) throws IOException {
+        final DurationTimer timer = LogUtil.startTimerIfDebugEnabled(LOGGER);
+        destZipArchiveEntry.setCompressedSize(sourceZipArchiveEntry.getCompressedSize());
+        destZipArchiveEntry.setCrc(sourceZipArchiveEntry.getCrc());
+        destZipArchiveEntry.setExternalAttributes(sourceZipArchiveEntry.getExternalAttributes());
+        destZipArchiveEntry.setExtra(sourceZipArchiveEntry.getExtra());
+        destZipArchiveEntry.setExtraFields(sourceZipArchiveEntry.getExtraFields());
+        destZipArchiveEntry.setGeneralPurposeBit(sourceZipArchiveEntry.getGeneralPurposeBit());
+        destZipArchiveEntry.setInternalAttributes(sourceZipArchiveEntry.getInternalAttributes());
+        destZipArchiveEntry.setMethod(sourceZipArchiveEntry.getMethod());
+        destZipArchiveEntry.setRawFlag(sourceZipArchiveEntry.getRawFlag());
+        destZipArchiveEntry.setSize(sourceZipArchiveEntry.getSize());
+
+        zipArchiveOutputStream.addRawArchiveEntry(sourceZipArchiveEntry, inputStream);
+        LOGGER.debug(() -> LogUtil.message(
+                "writeRawStream() - path: {}, zipArchiveEntry: {}, bytes: {}, duration: {}",
+                path,
+                destZipArchiveEntry,
+                ZipUtil.getEntryUncompressedSize(sourceZipArchiveEntry)
+                        .map(Objects::toString)
+                        .orElse("?"),
+                timer));
     }
 
     void putArchiveEntry(final ZipArchiveEntry zipArchiveEntry) throws IOException {

--- a/stroom-proxy/stroom-proxy-app/src/test/java/stroom/proxy/app/TestProxyConfigModule.java
+++ b/stroom-proxy/stroom-proxy-app/src/test/java/stroom/proxy/app/TestProxyConfigModule.java
@@ -52,9 +52,10 @@ class TestProxyConfigModule {
 
         final Predicate<Class<?>> classFilter = clazz ->
                 clazz.getSimpleName().endsWith("Config")
-                        && !clazz.equals(AbstractConfig.class)
-                        && !clazz.equals(ProxyConfig.class)
-                        && IsProxyConfig.class.isAssignableFrom(clazz);
+                && !clazz.equals(AbstractConfig.class)
+                && !clazz.equals(ProxyConfig.class)
+                && IsProxyConfig.class.isAssignableFrom(clazz)
+                && !clazz.getSimpleName().startsWith("Test");
 
         LOGGER.info("Finding all IsProxyConfig classes");
 
@@ -67,7 +68,7 @@ class TestProxyConfigModule {
                 .map((ClassInfo classInfo2) -> {
                     try {
                         return classInfo2.load();
-                    } catch (Exception e) {
+                    } catch (Throwable e) {
                         throw new RuntimeException(LogUtil.message(
                                 "Unable to load class {}", classInfo2), e);
                     }
@@ -156,8 +157,8 @@ class TestProxyConfigModule {
 
         if (!classesWithMultipleInstances.isEmpty()) {
             LOGGER.error("The following IsProxyConfig classes have a different injected instance to the " +
-                    "instance in the AppConfig tree.\n" +
-                    "You need to add Guice bindings for them in AppConfigModule");
+                         "instance in the AppConfig tree.\n" +
+                         "You need to add Guice bindings for them in AppConfigModule");
             classesWithMultipleInstances.stream()
                     .sorted(Comparator.comparing(Class::getName))
                     .forEach(clazz -> {

--- a/stroom-proxy/stroom-proxy-remote-api/src/main/java/stroom/proxy/feed/remote/GetFeedStatusRequest.java
+++ b/stroom-proxy/stroom-proxy-remote-api/src/main/java/stroom/proxy/feed/remote/GetFeedStatusRequest.java
@@ -61,11 +61,14 @@ public class GetFeedStatusRequest extends RemoteRequest {
     @Override
     public String toString() {
         StringBuilder builder = new StringBuilder();
-        builder.append("request ");
+        builder.append("feed status request - feed: '");
         builder.append(feedName);
+        builder.append("'");
         if (senderDn != null) {
             builder.append(" - ");
+            builder.append("'");
             builder.append(senderDn);
+            builder.append("'");
         }
         return builder.toString();
     }

--- a/stroom-proxy/stroom-proxy-repo/src/main/java/stroom/proxy/repo/FrequencyExecutor.java
+++ b/stroom-proxy/stroom-proxy-repo/src/main/java/stroom/proxy/repo/FrequencyExecutor.java
@@ -1,6 +1,7 @@
 package stroom.proxy.repo;
 
 import stroom.util.concurrent.UncheckedInterruptedException;
+import stroom.util.logging.DurationTimer;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
 import stroom.util.logging.LogUtil;
@@ -43,7 +44,9 @@ public class FrequencyExecutor implements Managed {
             try {
                 runnableSupplier.get().run();
             } catch (final UncheckedInterruptedException e) {
-                LOGGER.debug(e::getMessage, e);
+                // Swallow the exception to keep the scheduled executor running
+                LOGGER.debug("Frequency executor interrupted '{}' task: {}",
+                        threadNamePrefix, LogUtil.exceptionMessage(e), e);
             } catch (final RuntimeException e) {
                 // Swallow the exception to keep the scheduled executor running
                 LOGGER.error("Error running frequency executor '{}' task: {}",
@@ -56,8 +59,11 @@ public class FrequencyExecutor implements Managed {
 
     @Override
     public void stop() {
-        LOGGER.debug("Stopping frequency executor '{}', frequency: {}", threadNamePrefix, frequency);
+        LOGGER.info("Stopping frequency executor '{}', frequency: {}", threadNamePrefix, frequency);
+        final DurationTimer timer = DurationTimer.start();
         executorService.shutdownNow();
+        LOGGER.debug("Stopped frequency executor '{}', frequency: {}, duration: {}",
+                threadNamePrefix, frequency, timer);
     }
 
     @Override

--- a/stroom-proxy/stroom-proxy-repo/src/main/java/stroom/proxy/repo/ParallelExecutor.java
+++ b/stroom-proxy/stroom-proxy-repo/src/main/java/stroom/proxy/repo/ParallelExecutor.java
@@ -1,6 +1,7 @@
 package stroom.proxy.repo;
 
 import stroom.util.concurrent.UncheckedInterruptedException;
+import stroom.util.logging.DurationTimer;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
 import stroom.util.logging.LogUtil;
@@ -117,12 +118,13 @@ public class ParallelExecutor implements Managed {
                 // We need to release the blocked threads so the executor can shut down
                 resume();
             }
-            LOGGER.debug("Stopping parallel executor '{}', threadCount: {}",
+            LOGGER.info("Stopping parallel executor '{}', threadCount: {}",
                     threadNamePrefix, threadCount);
+            final DurationTimer timer = DurationTimer.start();
             executorService.shutdownNow();
             final boolean didTerminate = executorService.awaitTermination(1, TimeUnit.DAYS);
-            LOGGER.debug("Stopped parallel executor '{}', threadCount: {}, didTerminate: {}",
-                    threadNamePrefix, threadCount, didTerminate);
+            LOGGER.info("Stopped parallel executor '{}', threadCount: {}, didTerminate: {}, duration: {}",
+                    threadNamePrefix, threadCount, didTerminate, timer);
             isStopped.set(true);
         } else {
             throw new IllegalStateException(LogUtil.message(
@@ -197,6 +199,10 @@ public class ParallelExecutor implements Managed {
             LOGGER.debug("Running task");
             try {
                 task.run();
+            } catch (UncheckedInterruptedException e) {
+                // Swallow the exception to keep this thread running
+                LOGGER.debug("Parallel executor interrupted: '{}' task: {}",
+                        threadNamePrefix, LogUtil.exceptionMessage(e), e);
             } catch (Exception e) {
                 // Swallow the exception to keep this thread running
                 LOGGER.error("Error running parallel executor '{}' task: {}",

--- a/stroom-util/src/main/java/stroom/util/logging/LogUtil.java
+++ b/stroom-util/src/main/java/stroom/util/logging/LogUtil.java
@@ -5,6 +5,7 @@ import stroom.util.concurrent.DurationAdder;
 import stroom.util.shared.ModelStringUtil;
 
 import com.google.common.base.Strings;
+import org.slf4j.Logger;
 import org.slf4j.helpers.MessageFormatter;
 
 import java.io.PrintWriter;
@@ -427,6 +428,8 @@ public final class LogUtil {
     /**
      * Return the value supplied by supplier. Any exceptions will be swallowed and logged to debug only.
      * Useful for getting values for logging that may throw.
+     *
+     * @return The supplied value or an empty optional if an exception is thrown and swallowed.
      */
     public static <T> Optional<T> swallowExceptions(final Supplier<T> supplier) {
         try {
@@ -440,6 +443,8 @@ public final class LogUtil {
     /**
      * Return the value supplied by supplier. Any exceptions will be swallowed and logged to debug only.
      * Useful for getting values for logging that may throw.
+     *
+     * @return The supplied value or an empty optional if an exception is thrown and swallowed.
      */
     public static OptionalLong swallowExceptions(final LongSupplier supplier) {
         try {
@@ -453,6 +458,8 @@ public final class LogUtil {
     /**
      * Return the value supplied by supplier. Any exceptions will be swallowed and logged to debug only.
      * Useful for getting values for logging that may throw.
+     *
+     * @return The supplied value or an empty optional if an exception is thrown and swallowed.
      */
     public static OptionalInt swallowExceptions(final IntSupplier supplier) {
         try {
@@ -461,5 +468,27 @@ public final class LogUtil {
             LOGGER.debug("Error swallowed", e);
             return OptionalInt.empty();
         }
+    }
+
+    /**
+     * If DEBUG logging is enabled, create, start and return a new {@link DurationTimer}
+     * instance, else just return null.
+     * This avoids unnecessary object creation if DEBUG is not enabled.
+     */
+    public static DurationTimer startTimerIfDebugEnabled(final Logger logger) {
+        return logger != null && logger.isDebugEnabled()
+                ? DurationTimer.start()
+                : null;
+    }
+
+    /**
+     * If DEBUG logging is enabled, create, start and return a new {@link DurationTimer}
+     * instance, else just return null.
+     * This avoids unnecessary object creation if TRACE is not enabled.
+     */
+    public static DurationTimer startTimerIfTraceEnabled(final Logger logger) {
+        return logger != null && logger.isTraceEnabled()
+                ? DurationTimer.start()
+                : null;
     }
 }

--- a/stroom-util/src/main/java/stroom/util/zip/ZipUtil.java
+++ b/stroom-util/src/main/java/stroom/util/zip/ZipUtil.java
@@ -17,8 +17,10 @@
 package stroom.util.zip;
 
 import stroom.util.io.AbstractFileVisitor;
+import stroom.util.io.ByteSize;
 import stroom.util.io.StreamUtil;
 
+import org.apache.commons.compress.archivers.ArchiveEntry;
 import org.apache.commons.compress.archivers.zip.Zip64Mode;
 import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
 import org.apache.commons.compress.archivers.zip.ZipArchiveInputStream;
@@ -40,6 +42,7 @@ import java.nio.file.attribute.BasicFileAttributes;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
+import java.util.Optional;
 import java.util.function.Predicate;
 import java.util.regex.Pattern;
 
@@ -94,7 +97,7 @@ public final class ZipUtil {
     }
 
     public static void zip(final Path dir,
-                            final ZipArchiveOutputStream zipOutputStream) {
+                           final ZipArchiveOutputStream zipOutputStream) {
         zip(dir, zipOutputStream, path -> true, name -> true);
     }
 
@@ -182,5 +185,26 @@ public final class ZipUtil {
             }
         }
         return pathList;
+    }
+
+    /**
+     * Get the uncompressed size of the entry, if known.
+     */
+    public static Optional<ByteSize> getEntryUncompressedSize(final ArchiveEntry archiveEntry) {
+        return Optional.ofNullable(archiveEntry)
+                .map(ArchiveEntry::getSize)
+                .filter(size -> size != ArchiveEntry.SIZE_UNKNOWN)
+                .map(ByteSize::ofBytes);
+    }
+
+    /**
+     * @return True if the uncompressed size of the entry is known.
+     */
+    public static boolean hasKnownUncompressedSize(final ArchiveEntry archiveEntry) {
+        if (archiveEntry == null) {
+            return false;
+        } else {
+            return archiveEntry.getSize() != ArchiveEntry.SIZE_UNKNOWN;
+        }
     }
 }

--- a/unreleased_changes/20250415_163812_915__4863.md
+++ b/unreleased_changes/20250415_163812_915__4863.md
@@ -1,0 +1,24 @@
+* Issue **#4863** : Stop errors being logged on proxy shutdown.
+
+
+```sh
+# ********************************************************************************
+# Issue title: Proxy logging stack traces when shutting down
+# Issue link:  https://github.com/gchq/stroom/issues/4863
+# ********************************************************************************
+
+# ONLY the top line will be included as a change entry in the CHANGELOG.
+# The entry should be in GitHub flavour markdown and should be written on a SINGLE
+# line with no hard breaks. You can have multiple change files for a single GitHub issue.
+# The  entry should be written in the imperative mood, i.e. 'Fix nasty bug' rather than
+# 'Fixed nasty bug'.
+#
+# Examples of acceptable entries are:
+#
+#
+# * Issue **123** : Fix bug with an associated GitHub issue in this repository
+#
+# * Issue **namespace/other-repo#456** : Fix bug with an associated GitHub issue in another repository
+#
+# * Fix bug with no associated GitHub issue.
+```

--- a/unreleased_changes/20250415_171225_150__4879.md
+++ b/unreleased_changes/20250415_171225_150__4879.md
@@ -1,0 +1,24 @@
+* Issue **#4879** : Proxy - Improve concurrency protection when closing aggregates.
+
+
+```sh
+# ********************************************************************************
+# Issue title: Proxy - exception when closing aggregates
+# Issue link:  https://github.com/gchq/stroom/issues/4879
+# ********************************************************************************
+
+# ONLY the top line will be included as a change entry in the CHANGELOG.
+# The entry should be in GitHub flavour markdown and should be written on a SINGLE
+# line with no hard breaks. You can have multiple change files for a single GitHub issue.
+# The  entry should be written in the imperative mood, i.e. 'Fix nasty bug' rather than
+# 'Fixed nasty bug'.
+#
+# Examples of acceptable entries are:
+#
+#
+# * Issue **123** : Fix bug with an associated GitHub issue in this repository
+#
+# * Issue **namespace/other-repo#456** : Fix bug with an associated GitHub issue in another repository
+#
+# * Fix bug with no associated GitHub issue.
+```

